File: lib/rgfa/line/segment/length_gfa1.rb
module RGFA::Line::Segment::LengthGFA1

  # @!macro [new] length
  #   @return [Integer] value of LN tag, if segment has LN tag
  #   @return [Integer] sequence length if no LN and sequence not "*"
  # @return [nil] if sequence is "*"
  # @see #length!
  def length
    if self.LN
      self.LN
    elsif !sequence.placeholder? and !sequence.kind_of?(RGFA::Placeholder)
      sequence.length
    else
      nil
    end
  end

  # @!macro length
  # @!macro [new] length_needed
  #   @raise [RGFA::NotFoundError] if not an LN tag and
  #     the sequence is "*"
  # @see #length
  def length!
    l = self.length()
    raise RGFA::NotFoundError,
      "No length information available" if l.nil?
    return l
  end

  # @raise [RGFA::InconsistencyError]
  #    if sequence length and LN tag are not consistent.
  def validate_length!
    if !sequence.placeholder? and tagnames.include?(:LN)
      if self.LN != sequence.length
        raise RGFA::InconsistencyError,
          "Length in LN tag (#{self.LN}) "+
          "is different from length of sequence field (#{sequence.length})"
      end
    end
  end

  private

  def validate_record_type_specific_info!
    validate_length!
  end

end

File: gfapy/line/segment/length_gfa1.py
import gfapy

class LengthGFA1:

  @property
  def length(self):
    """
    Returns
    -------
    int
      Value of LN tag, if segment has LN tag.
    int
      Sequence length if no LN and sequence not "*".
    None
      If sequence is "*".

    See Also
    --------
    try_get_length
    """
    if self.LN:
      return self.LN
    elif not gfapy.is_placeholder(self.sequence):
      return len(self.sequence)
    else:
      return None

  def try_get_length(self):
    """
    Raises
    ------
    gfapy.NotFoundError
      If not an LN tag and the sequence is "*".

    See Also
    --------
    __len__
    """
    l = self.length
    if l is None:
      raise gfapy.NotFoundError("No length information available")
    return l

  def validate_length(self):
    """
    Raises
    ------
    gfapy.InconsistencyError
      If sequence length and LN tag are not consistent.
    """
    if not gfapy.is_placeholder(self.sequence) and "LN" in self.tagnames:
      if self.LN != len(self.sequence):
        raise gfapy.InconsistencyError(
          "Segment: {}\n".format(str(self))+
          "Length in LN tag ({}) ".format(self.LN)+
          "is different from length of sequence field ({})"
          .format(len(self.sequence)))

  def _validate_record_type_specific_info(self):
    self.validate_length()
